tools/src/org.graalvm.tools.lsp/src/org/graalvm/tools/lsp/server/utils/TextMap.java

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright (c) 2013, 2016, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package org.graalvm.tools.lsp.server.utils;
+
+import java.util.ArrayList;
+
+/**
+ * A utility for converting between coordinate systems in a string of text interspersed with newline
+ * characters. The coordinate systems are:
+ * <ul>
+ * <li>0-based character offset from the beginning of the text, where newline characters count as a
+ * single character and the first character in the text occupies position 0.</li>
+ * <li>1-based position in the 2D space of lines and columns, in which the first position in the
+ * text is at (1,1).</li>
+ * </ul>
+ * <p>
+ * This utility is based on positions occupied by characters, not text stream positions as in a text
+ * editor. The distinction shows up in editors where you can put the cursor just past the last
+ * character in a buffer; this is necessary, among other reasons, so that you can put the edit
+ * cursor in a new (empty) buffer. For the purposes of this utility, however, there are no character
+ * positions in an empty text string and there are no lines in an empty text string.
+ * <p>
+ * A newline character designates the end of a line and occupies a column position.
+ * <p>
+ * If the text ends with a character other than a newline, then the characters following the final
+ * newline character count as a line, even though not newline-terminated.
+ * <p>
+ * <strong>Limitations:</strong>
+ * <ul>
+ * <li>Does not handle multiple character encodings correctly.</li>
+ * <li>Treats tabs as occupying 1 column.</li>
+ * <li>Does not handle multiple-character line termination sequences correctly.</li>
+ * </ul>
+ */
+final class TextMap {
```

Comment:
we should not copy the TextMap implementation. Or do you want to write tests for it?
Can we use the Source capabilities instead? (Even if it is not evaluated in the end)